File: Swift/BasicExample/ChickenDelivery15686/ChickenDelivery15686/main.swift
//
//  main.swift
//  ChickenDelivery15686
//
//  Created by 조윤영 on 2020/12/31.
//

import Foundation

var N: Int = 0
var M: Int = 0
var map = [[Int]]()
var houses = [(y: Int, x: Int)]()
var chickens = [(y: Int, x: Int)]()
var order = [Int]()
func input() {
    let inputList: [Int] = readLine()!.split(separator: " ").map{ Int($0)! }
    N = inputList[0]
    M = inputList[1]
    map = [[Int]](repeating: [Int](repeating: 0, count: N), count: N)
    for i in 0..<N {
        let mapInputList: [Int] = readLine()!.split(separator: " ").map{ Int($0)! }
        for j in 0..<N {
            if mapInputList[j] == 1 { houses.append((y: i,x:  j)) }
            else if mapInputList[j] == 2 { chickens.append((y: i, x:  j)) }
        }
    }
}

func getChickenDist() -> Int {
    var result = 0
    houses.forEach { house in
        var dist = Int.max
        order.forEach { o in
            let diffY = abs(chickens[o].y - house.y)
            let dffX = abs(chickens[o].x - house.x)
            let diffSum = diffY + dffX
            dist = dist > diffSum ? diffSum : dist
        }
        result += dist
    }
    
    return result
}

var answer: Int = Int.max
func DFS(_ cnt: Int) {
    if cnt == M {
        answer = min(answer, getChickenDist())
        return
    }
    
    var idx = order.count == 0 ? 0 : order.last! + 1

    while idx < chickens.count {
        order.append(idx)
        DFS(cnt+1)
        order.removeLast()
        idx += 1
    }
}

func solution() {
    DFS(0)
    print(answer)
}

input()
solution()

File: C++/BasicExample/ChickenDelivery15686/ChickenDelivery15686/main.cpp
//
//  main.cpp
//  ChickenDelivery15686
//
//  Created by 조윤영 on 2021/01/09.
//
#define INF 1e9
#define MAX_SIZE 51
#include <iostream>
#include <vector>
#include <queue>
#include <cmath>

using namespace std;
int N, M = 0;
int map[MAX_SIZE][MAX_SIZE];
int visited[14];
struct Pos {
    int y;
    int x;
    Pos(int _y, int _x) {
        y = _y;
        x = _x;
    };
};

struct CostInfo {
    int cost;
    int cnt;
    int index;
    CostInfo(int _cost, int _cnt, int _index) {
        cost = _cost;
        cnt = _cnt;
        index = _index;
    };
};
vector<Pos> houses;
vector<Pos> chickens;
vector<Pos> selected_chickens;
int answer = INF;

void input() {
    cin>> N >> M;
    for(int i=0; i<N; i++) {
        for(int j=0; j<N; j++) {
            cin>> map[i][j];
            if(map[i][j] == 1) houses.push_back(Pos(i, j));
            else if(map[i][j] == 2) chickens.push_back(Pos(i, j));
        }
    }
}

int calculateCost() {
    int result = 0;
    
    for(int i=0; i<houses.size(); i++) {
        int minCost = INF;
        for(int j=0; j<selected_chickens.size(); j++) {
            int cost = abs((houses[i].y - selected_chickens[j].y)) + abs((houses[i].x - selected_chickens[j].x));
            minCost = min(minCost, cost);
        }
        
        result += minCost;
    }
    return result;
}

//조합 문제
void dfs(int index, int cnt) {
    if(cnt == M) {
        answer  = min(answer, calculateCost());
        return;
    }
    
    for(int i=index; i<chickens.size(); i++) {
        if(visited[i]) continue;
        visited[i] = true;
        selected_chickens.push_back(chickens[i]);
        dfs(i+1, cnt+1);
        visited[i] = false;
        selected_chickens.pop_back();
    }
}
void solution() {
    dfs(0,0);
    cout<<answer<<"\n";
}

int main(int argc, const char * argv[]) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    input();
    solution();
    return 0;
}
